design/design-generic-references.md

```diff
@@ -0,0 +1,211 @@
+  of an RDS Instance.
+* Instead of hard-coding region for all managed resources, you may want to refer
+  to a `ConfigMap` or a `Secret` and let the managed resource take that
+  information from there.
+* In Composition, you can use values of the fields only between the children of
+  the same composite instance, but there are cases where you'd like to use a
+  separate resource that is not bound to any composite instance to store the
+  information and serve all composites.
+
+In order to enable such use cases, we need a generic version of the cross-resource
+references where you can specify all metadata required to find the referenced
+resource as well as details about how to extract the value you need and where to
+put it in the referencer object.
+
+## Proposal
+
+All managed resources will have a top-level spec field that lists the generic
+references. They will be resolved in order and failure of any will cause
+reconciliation to fail. Inspired from the [latest developments in API conventions](https://github.com/kubernetes/community/pull/5748)
+regarding multi-kind references, the API will look like the following:
+```yaml
+# A generic reference to another managed resource.
+spec:
+  externalValues:
+  - fromObject:
+      group: ec2.aws.crossplane.io
+      version: v1alpha1
+      resource: vpcs
+      name: main-vpc
+      fieldPath: spec.forProvider.cidrBlock
+    toFieldPath: spec.cidrBlock
+```
+```yaml
+# A generic reference to a namespaced Kubernetes resource.
+spec:
+  externalValues:
+  - fromObject:
+      version: v1
+      resource: configmaps
+      name: common-settings
+      namespace: crossplane-system
+      fieldPath: data.region
+    toFieldPath: spec.forProvider.region
+```
+
+The syntax of `fieldPath` and `toFieldPath` fields will be similar to what we
+use in Composition, in line with [Kubernetes API Conventions](https://github.com/kubernetes/community/blob/744e270/contributors/devel/sig-architecture/api-conventions.md#selecting-fields).
+
+The provider pods will resolve the references via managed reconciler, hence they
+will need at least read-only permissions for any resource that's been referenced
+by an instance. Currently, every provider [has read access](https://github.com/crossplane/crossplane/blob/d8f57a8/internal/controller/rbac/provider/roles/roles.go#L66)
+to all of its own kinds, `Secret`s, `ConfigMap`s, `Lease`s and `Event`s in all
+namespaces. In order to reference resources other than the ones listed, users will
+have to create necessary RBAC resources manually to grant the permissions.
+
+### Implementation
+
+The new field will exist in all managed resources and the cross-resource reference
+fields will stay as is and we'll continue adding them wherever possible. They will
+still be the main way to refer to another resource. The generic references will
+help in cases where that's not possible or not implemented yet.
+
+The API struct and the generic resolver will live in Crossplane Runtime and be
```

Comment:
I think the previous version made it clear that references behaved like "required patches" and were resolved by core crossplane and the provider. Is it now only in the provider? I wonder how this will work if the managed resource is not valid before resolving the reference (schema/webhook).

I wonder if partially creating the managed resources (those without any unresolved dependencies) will be required in some cases.

Reply:
I think it's helpful to think of this closer to cross-resource references like `bucketRef` on the managed resources rather than patches on composition. Today, you can create a managed resource with `bucketRef` whose name doesn't correspond to anything. You'll see an error in the status but the resource creation will go through. The generic references in the managed resources will work the same way in the sense that you'll be able to create the resource and provider pod will try to resolve all references, both cross-resource and generic refs, and only after doing that successfully it will start calling external client functions, `Connect`, `Observe` etc.